sos/collector/clusters/juju.py

```diff
@@ -0,0 +1,214 @@
+                        node = f"{self.model_name}:{unit_info['machine']}"
+                        for sub_key, sub_value in unit_info.get(
+                            "subordinates", {}
+                        ).items():
+                            if sub_key.startswith(app + "/"):
+                                self.units[sub_key] = [node]
+
+        def add_machines(self, juju_status):
+            """Add machines to index.
+
+            If model does not have any applications it needs to be
+            manually added.
+            """
+            for machine in juju_status["machines"].keys():
+                node = f"{self.model_name}:{machine}"
+                self.machines[machine] = [node]
+
+    return Index(model_name)
+
+
+class juju(Cluster):
+    """
+    The juju cluster profile is intended to be used on juju managed clouds.
+    It"s assumed that `juju` is installed on the machine where `sos` is called,
+    and that the juju user has superuser privilege to the current controller.
+
+    By default, the sos reports will be collected from all the applications in
+    the current model. If necessary, you can filter the nodes by models /
+    applications / units / machines with cluster options.
+
+    Example:
+
+    sos collect --cluster-type juju -c "juju.models=sos" -c "juju.apps=a,b,c"
+
+    """
+
+    cmd = "juju"
+    cluster_name = "Juju Managed Clouds"
+    option_list = [
+        ("apps", "", "Filter node list by apps (comma separated regex)."),
+        ("units", "", "Filter node list by units (comma separated string)."),
+        ("models", "", "Filter node list by models (comma separated string)."),
+        (
+            "machines",
+            "",
+            "Filter node list by machines (comma separated string).",
+        ),
+    ]
+
+    def _cleanup_juju_output(self, output):
+        """Remove leading characters before {."""
+        return re.sub(r"(^[^{]*)(.*)", "\\2", output, 0, re.MULTILINE)
+
+    def _get_model_info(self, model_name):
+        """Parse juju status output and return target dict.
+
+        Here are couple helper functions to parse the juju principals units,
+        subordinate units and machines.
+        """
+        juju_status = self._execute_juju_status(model_name)
+
+        index = _get_index(model_name=model_name)
+        index.add_principals(juju_status)
+        index.add_subordinates(juju_status)
+        index.add_machines(juju_status)
+
+        return index
+
+    def _execute_juju_status(self, model_name):
+        model_option = f"-m {model_name}" if model_name else ""
+        format_option = "--format json"
+        status_cmd = f"{self.cmd} status {model_option} {format_option}"
+        res = self.exec_primary_cmd(status_cmd)
+        if not res["status"] == 0:
+            raise Exception(f"'{status_cmd}' returned error: {res['status']}")
+        juju_json_output = self._cleanup_juju_output((res["output"]))
+
+        juju_status = None
```

Comment:
I *think* this line is redundant (either the command in `try` succeeds and assigns a value to the variable, or we raise an exception and no return value matters?). But let be sure..